geode-core/src/main/java/org/apache/geode/cache/client/internal/QueueManagerImpl.java

```diff
@@ -853,7 +853,8 @@ void recoverPrimary(Set<ServerLocation> excludedServers) {
       return;
     }
     final boolean isDebugEnabled = logger.isDebugEnabled();
-    if (queueConnections.getPrimary() != null && !queueConnections.getPrimary().isDestroyed()) {
+    if (queueConnections != null && queueConnections.getPrimary() != null
+        && !queueConnections.getPrimary().isDestroyed()) {
```

Comment:
Under what circumstance could queueConnections become null?
If there is a condition for that, I would write a test case to verify the fix.

Reply:
There is the possibility for the `queueConnections` to be null during teardown of `CacheClientNotifierDUnitTest`. While the cache is closing and everything is being freed, the `RedundancySatisfierTask` continues to check `queueConnections` whenever it is referenced (because it is volatile). So if the `RedundancySatisfierTask` thread is slow to close and is still checking the `queueConnections` after it has been freed, it could cause an NPE.

Do you think that a test for that situation would be valuable? I worry that it will just be forcing weird timing, and would probably be flaky.

Reply:
Ok I think I did what you described, but please let me know if I misunderstood what you were suggesting. I think it should be safe to remove the null check that I previously added, but since it's hard to test I'm not sure. Thoughts?

Reply:
Yes, you did what I described. But, unfortunately, I have realized that this will not work (see the test cases failing) because the `queueConnections` is a copy on write structure, meaning that every time it is updated, a fresh copy is created.
With this change, the `RedundancySatisfierTask` will eventually not point to the same instance as the `QueueManagerImpl` instance.

The problem with the null check added is that I think it is just one of the places where it should be added. Plus, you would always be in risk on hitting the race condition even if you first checked that the `queueConnections` is null. What if right after you check it is set to null?

How about holding a reference to the `QueueManagerImpl` instance in the `RedundancySatisfierTask` instance?
You could pass it when you create the `RedundancySatisfierTask` as follows:

`redundancySatisfierTask = new RedundancySatisfierTask(this);`

Reply:
I feel "redundancySatisfierTask = new RedundancySatisfierTask(this);" could be a good idea. 

There're so many places are accessing queueConnections directly without checking if it's null.